Refactor robotest configuration (automatically keep upgrade_from releases current)

wadells opened the pull request:
## Description
This changeset introduces 4 notable changes:

1. Robotest upgrade_from gravity releases are dynamically determined. No more manually bumping patch releases!
2. I prune one dimension (cluster size) from nightly upgrade testing.  We now have a single vector though the size dimension, instead of multiplying it by OS and version. This cuts 20 upgrade tests as seen in the diffs in Testing Done.
3. I refactored our scripts such that shared code between nightly & pr is no longer duplicated, per [a-palchikov's request](https://github.com/gravitational/gravity/pull/1761#discussion_r444845784).  This will also help limit merge conflicts to `pr_config.sh` and `nightly_config.sh` as the other files should be largely consistent between branches.
4. We no longer use the default `tele` 3.49 in jenkin's path to pull artifacts.

## Type of change
* Internal change (not necessarily a bug fix or a new feature)

## Linked tickets and other PRs
Fixes #1199.
Fixes #1154.
This will catch issues like #1697.

Needs ports to all active releases.

## TODOs
- [x] Self-review the change
- [x] Perform manual testing
- [x] Write documentation
- [x] Address review feedback

## Implementation
See the [comment in utils.sh](https://github.com/gravitational/gravity/pull/1837/files#diff-ab0ab57e4b51b95f4e7b0d9ed899403bR35-R57) for reasoning behind how recommended upgrade_from releases are calculated.

We have a lot of release edge cases in the wild:

<details><summary>gravity release edge cases</summary>

* 6.1.23 doesn't exist.
* The following git tags don't have entries in the enterprise get.gravitational.io application catalog:
```
5.2.0-alpha.2
5.3.0-alpha.1
5.4.0-alpha.1
5.5.0-alpha.1
5.5.0-alpha.2
5.5.11
5.5.16
5.5.25
5.5.29
5.5.39
5.6.2
5.6.3
6.0.0-alpha.1
6.0.0-rc.3
6.2.0-alpha.1
7.0.0-alpha.1
7.1.0-alpha.1
```
* Going the other way, the following are published to get.gravitational.io, but have no git tag:
```
5.2.0-rc.2.2
5.3.3-3
5.5.40-dev.6
5.5.46-dev.5
5.5.49-dev.11
6.1.28-dev.4
7.0.0-rc.2.3
```
</details>

This PR necessarily introduces sources non-determinism unrelated to the PR being tested -- git tags on other branches and published artifacts both can change under an extant PRs feet. As such, I've been pretty pretty aggressive in accommodating release mistakes and edge cases instead of having robotest fail and pushing triaging/fixing release inconsistencies onto bystanders who happened to submit a PR at an unfortunate time.

## Testing done
Most importantly, here are try builds showing the new PR suite executes in the same manner as the old code:

* https://jenkins.gravitational.io/job/gravity-try-build/53/ (failed with a shrink issue unrelated to these changes)
* https://jenkins.gravitational.io/job/gravity-try-build/54/ (success)

Next, here are the new suites.  If you have opinions about if we should be testing more/less, these are probably the most interesting thing to critique.
<details><summary>./build.assets/robotest/nightly_config.sh</summary>

```
walt@work:~/git/gravity$ ./build.assets/robotest/nightly_config.sh       
install={"flavor":"three","nodes":3,"role":"node","os":"redhat:7","storage_driver":"overlay2"}
install={"flavor":"six","nodes":6,"role":"node","os":"redhat:7","storage_driver":"overlay2"}
install={"flavor":"three","nodes":3,"role":"node","os":"debian:9","storage_driver":"overlay2"}
install={"flavor":"six","nodes":6,"role":"node","os":"debian:9","storage_driver":"overlay2"}
install={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:16","storage_driver":"overlay2"}
install={"flavor":"six","nodes":6,"role":"node","os":"ubuntu:16","storage_driver":"overlay2"}
install={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:18","storage_driver":"overlay2"}
install={"flavor":"six","nodes":6,"role":"node","os":"ubuntu:18","storage_driver":"overlay2"}
install={"installer_url":"/installer/opscenter.tar","nodes":1,"flavor":"standalone","role":"node","os":"ubuntu:18","ops_advertise_addr":"example.com:443"}
resize={"to":3,"flavor":"one","nodes":1,"role":"node","state_dir":"/var/lib/telekube","os":"ubuntu:18","storage_driver":"overlay2"}
resize={"to":6,"flavor":"three","nodes":3,"role":"node","state_dir":"/var/lib/telekube","os":"ubuntu:18","storage_driver":"overlay2"}
shrink={"nodes":3,"flavor":"three","role":"node","os":"redhat:7"}
upgrade={"flavor":"three","nodes":3,"role":"node","os":"redhat:7","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"flavor":"six","nodes":6,"role":"node","os":"redhat:7","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"flavor":"one","nodes":1,"role":"node","os":"redhat:7","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:16","from":"/telekube_7.0.0.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:16","from":"/telekube_6.2.0.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"flavor":"three","nodes":3,"role":"node","os":"redhat:7","from":"/telekube_6.2.5.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"flavor":"three","nodes":3,"role":"node","os":"redhat:7","from":"/telekube_6.3.18.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"flavor":"three","nodes":3,"role":"node","os":"centos:7","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"flavor":"three","nodes":3,"role":"node","os":"debian:9","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:18","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
```
</details>

<details><summary>./build.assets/robotest/pr_config.sh</summary>

```
walt@work:~/git/gravity$ ./build.assets/robotest/pr_config.sh     
install={"flavor":"three","nodes":3,"role":"node","os":"redhat:7","storage_driver":"overlay2"}
install={"installer_url":"/installer/opscenter.tar","nodes":1,"flavor":"standalone","role":"node","os":"ubuntu:18","ops_advertise_addr":"example.com:443"}
resize={"to":3,"flavor":"one","nodes":1,"role":"node","state_dir":"/var/lib/telekube","os":"ubuntu:18","storage_driver":"overlay2"}
shrink={"nodes":3,"flavor":"three","role":"node","os":"redhat:7"}
upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:16","from":"/telekube_7.0.0.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:18","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
```
</details>

Or, perhaps more usefully, here are the diffs between the old config and the new

<details><summary>pr differences (+0 tests total)</summary>

```
walt@work:~/git/gravity$ wc -l pr.master.txt pr.dynamic.txt 
   6 pr.master.txt
   8 pr.dynamic.txt
  14 total
walt@work:~/git/gravity$ diff pr.master.txt pr.dynamic.txt
5,6c5,8
< upgrade={"flavor":"three","nodes":3,"role":"node","service_uid":997,"service_gid":994,"os":"ubuntu:16","storage_driver":"overlay2","from":"/telekube_7.0.0.tar"}
< upgrade={"flavor":"three","nodes":3,"role":"node","service_uid":997,"service_gid":994,"os":"ubuntu:18","storage_driver":"overlay2","from":"/telekube_7.0.10.tar"}
---
> upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:16","from":"/telekube_7.0.0.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
> upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:18","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
```
</details>

<details><summary>nightly differences (-20 tests total) </summary>

```
walt@work:~/git/gravity$ wc -l nightly.master.txt nightly.dynamic.txt
  42 nightly.master.txt
  22 nightly.dynamic.txt
  64 total
walt@work:~/git/gravity$ diff nightly.master.txt nightly.dynamic.txt                                                                                                                          
13,42c13,22                                                                                                                                                                                   
< upgrade={"flavor":"one","nodes":1,"role":"node","os":"centos:7","storage_driver":"overlay2","from":"/telekube_6.2.5.tar"}                                                                   
< upgrade={"flavor":"one","nodes":1,"role":"node","os":"centos:7","storage_driver":"overlay2","from":"/telekube_6.3.18.tar"}                                                                  
< upgrade={"flavor":"one","nodes":1,"role":"node","os":"centos:7","storage_driver":"overlay2","from":"/telekube_7.0.10.tar"}                                                                  
< upgrade={"flavor":"one","nodes":1,"role":"node","os":"debian:9","storage_driver":"overlay2","from":"/telekube_6.2.5.tar"}                                                                   
< upgrade={"flavor":"one","nodes":1,"role":"node","os":"debian:9","storage_driver":"overlay2","from":"/telekube_6.3.18.tar"}                                                                  
< upgrade={"flavor":"one","nodes":1,"role":"node","os":"debian:9","storage_driver":"overlay2","from":"/telekube_7.0.10.tar"}                                                                  
< upgrade={"flavor":"one","nodes":1,"role":"node","os":"ubuntu:16","storage_driver":"overlay2","from":"/telekube_7.0.0.tar"}                                                                  
< upgrade={"flavor":"one","nodes":1,"role":"node","os":"ubuntu:18","storage_driver":"overlay2","from":"/telekube_6.2.5.tar"}                                                                  
< upgrade={"flavor":"one","nodes":1,"role":"node","os":"ubuntu:18","storage_driver":"overlay2","from":"/telekube_6.3.18.tar"}                                                                 
< upgrade={"flavor":"one","nodes":1,"role":"node","os":"ubuntu:18","storage_driver":"overlay2","from":"/telekube_7.0.10.tar"}                                                                 
< upgrade={"flavor":"six","nodes":6,"role":"node","os":"centos:7","storage_driver":"overlay2","from":"/telekube_6.2.5.tar"}                                                                   
< upgrade={"flavor":"six","nodes":6,"role":"node","os":"centos:7","storage_driver":"overlay2","from":"/telekube_6.3.18.tar"}                                                                  
< upgrade={"flavor":"six","nodes":6,"role":"node","os":"centos:7","storage_driver":"overlay2","from":"/telekube_7.0.10.tar"}                                                                  
< upgrade={"flavor":"six","nodes":6,"role":"node","os":"debian:9","storage_driver":"overlay2","from":"/telekube_6.2.5.tar"}                                                                   
< upgrade={"flavor":"six","nodes":6,"role":"node","os":"debian:9","storage_driver":"overlay2","from":"/telekube_6.3.18.tar"}                                                                  
< upgrade={"flavor":"six","nodes":6,"role":"node","os":"debian:9","storage_driver":"overlay2","from":"/telekube_7.0.10.tar"}                                                                  
< upgrade={"flavor":"six","nodes":6,"role":"node","os":"ubuntu:16","storage_driver":"overlay2","from":"/telekube_7.0.0.tar"}                                                                  
< upgrade={"flavor":"six","nodes":6,"role":"node","os":"ubuntu:18","storage_driver":"overlay2","from":"/telekube_6.2.5.tar"}                                                                  
< upgrade={"flavor":"six","nodes":6,"role":"node","os":"ubuntu:18","storage_driver":"overlay2","from":"/telekube_6.3.18.tar"}                                                                 
< upgrade={"flavor":"six","nodes":6,"role":"node","os":"ubuntu:18","storage_driver":"overlay2","from":"/telekube_7.0.10.tar"}
< upgrade={"flavor":"three","nodes":3,"role":"node","os":"centos:7","storage_driver":"overlay2","from":"/telekube_6.2.5.tar"}
< upgrade={"flavor":"three","nodes":3,"role":"node","os":"centos:7","storage_driver":"overlay2","from":"/telekube_6.3.18.tar"}
< upgrade={"flavor":"three","nodes":3,"role":"node","os":"centos:7","storage_driver":"overlay2","from":"/telekube_7.0.10.tar"}
< upgrade={"flavor":"three","nodes":3,"role":"node","os":"debian:9","storage_driver":"overlay2","from":"/telekube_6.2.5.tar"}
< upgrade={"flavor":"three","nodes":3,"role":"node","os":"debian:9","storage_driver":"overlay2","from":"/telekube_6.3.18.tar"}
< upgrade={"flavor":"three","nodes":3,"role":"node","os":"debian:9","storage_driver":"overlay2","from":"/telekube_7.0.10.tar"}
< upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:16","storage_driver":"overlay2","from":"/telekube_7.0.0.tar"}
< upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:18","storage_driver":"overlay2","from":"/telekube_6.2.5.tar"}
< upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:18","storage_driver":"overlay2","from":"/telekube_6.3.18.tar"}
< upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:18","storage_driver":"overlay2","from":"/telekube_7.0.10.tar"}
---
> upgrade={"flavor":"one","nodes":1,"role":"node","os":"redhat:7","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
> upgrade={"flavor":"six","nodes":6,"role":"node","os":"redhat:7","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
> upgrade={"flavor":"three","nodes":3,"role":"node","os":"centos:7","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
> upgrade={"flavor":"three","nodes":3,"role":"node","os":"debian:9","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
> upgrade={"flavor":"three","nodes":3,"role":"node","os":"redhat:7","from":"/telekube_6.2.5.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
> upgrade={"flavor":"three","nodes":3,"role":"node","os":"redhat:7","from":"/telekube_6.3.18.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
> upgrade={"flavor":"three","nodes":3,"role":"node","os":"redhat:7","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
> upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:16","from":"/telekube_6.2.0.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
> upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:16","from":"/telekube_7.0.0.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
> upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:18","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
```
</details>

## Additional information
This is round 2 of the work started at #1761

wadells committed:
Refactor robotest scripts, keep upgrades current.

This introduces two significant changes:

1) As suggested by @a-palchikov here:

  https://github.com/gravitational/gravity/pull/1761#discussion_r444845784

Functionality common to nightly and pr runs are is now shared.  The
configuration differences can be generated independently which will test
future changes to either configuration.

2) Robotest "upgrade from" versions are now always kept up to date,
using the latest tag on the branch.

This alleviates maintainers of the burden of bumping release versions
again and again.

The current git tag strategy comes from discussion here:

  https://github.com/gravitational/gravity/pull/1761#discussion_r444994839

Contributes to #1199. Intended to catch issues like #1691.

wadells committed:
Remove DOCKER_STORAGE_DRIVERS from nightly script.

Gravity only supports overlay2.  No need to keep the abstraction around
when we only test with one driver.

wadells committed:
Remove size from multiplicative upgrade matrix.

We were running nearly 40 upgrade tests -- a huge usage of cloud
resources -- because we did a size x OS x version matrix.  I've reduced
this to OS x version, with a single vector through different sizes.
redhat:7 and 3 node arbitrarily chosen for the size vector.

wadells committed:
Remove unused WAIT_FOR_INSTALLER.

This is never mentioned in the entire source history of robotest.
I think it came from some discarded work in PR #90.

wadells committed:
Add a robotest README.md.

This will help future maintainers understand why files are organized the
way they are.

wadells committed:
Use tele from the build to fetch artifacts.

Using the default tele in the path means our build is dependent on
implicit unmanaged state on the jenkins builder.  Plus, using the
tele from the build means we test fetching artifacts with tele every
time we run robotest.

Fixes #1154.

wadells commented on build.assets/robotest/nightly_config.sh:
```diff
@@ -0,0 +1,101 @@
+UPGRADE_VERSIONS=${!UPGRADE_MAP[@]}
+
+function build_upgrade_size_suite {
+  local from_tarball=/$(semver_to_tarball $(latest_release $(branch 7.0.x)))
```
`recommended_upgrade_tag` again.  Possibly unset latest_release if bash has working closures.

wadells replied:
Fixed in ab36936.

wadells commented on build.assets/robotest/nightly_config.sh:
```diff
@@ -0,0 +1,101 @@
+# UPGRADE_MAP maps gravity version -> list of linux distros to upgrade from
+declare -A UPGRADE_MAP
+
+UPGRADE_MAP[$(latest_release $(branch 7.0.x))]="centos:7 debian:9 ubuntu:18" # compatible LTS version
+UPGRADE_MAP[7.0.0]="ubuntu:16"
+UPGRADE_MAP[$(latest_release $(branch 6.3.x))]="redhat:7" # compatible non-LTS version
+# UPGRADE_MAP[6.3.0]="ubuntu:16"  # disabled due to https://github.com/gravitational/gravity/issues/1009
+UPGRADE_MAP[$(latest_release $(branch 6.2.x))]="redhat:7" # compatible non-LTS version
```
These should all use `recommended_upgrade_tag` instead of `latest_release` to make sure this still does the right thing when it is several months old. and 7.0.x may have etcds that are more current than this commit supports.

wadells replied:
Fixed in ab36936.

wadells committed:
Change latest_release -> recommended_release_tag in nightly.

Using latest_release was as mistake.

wadells committed:
Rename semver_to_tarball to tag_to_tarball.

We were using too many different terms for the same thing: semver,
release, and tag.  Cutting semver reduces it down to tag (needed because
we are working in git tags) and release.

r0mant commented on build.assets/robotest/README.md:
```diff
@@ -0,0 +1,19 @@
+* `run.sh` provides the glue between the top level Makefile, what robotest expects,
+and the dynamically generated configurations.
+* `*_config.sh` contain branch specific suite configurations.
+* `utils.sh` contains functions common to both pull requst (pr) and nightly
```
requst -> request

wadells replied:
Fixed in 257f484.

r0mant commented on build.assets/robotest/pr_config.sh:
```diff
@@ -0,0 +1,70 @@
+
+# UPGRADE_MAP maps gravity version -> list of linux distros to upgrade from
+declare -A UPGRADE_MAP
+UPGRADE_MAP[$(recommended_upgrade_tag $(branch 7.0.x))]="centos:7 debian:9 ubuntu:18" # compatible LTS version
```
Did we execute all 3 OS'es on each PR before as well, or is that a new thing?

wadells replied:
This was an unintentional change resulting from cut & paste from nightly at some point during the refactoring..  I will revert back to only ubuntu:18.

I think we probably test some additional OSes on PR, but I'd recommend those as installs / wrapped into expand/shrink instead of as upgrade (which is typically our heaviest scenario).

wadells replied:
Removed in a89b054.

r0mant commented on build.assets/robotest/test.sh:
```diff
@@ -0,0 +1,23 @@
+fi
+
+# check a more current 7.0.x has a more current corresponding 6.0.x
+if [[ "$(recommended_upgrade_tag_between $(branch 6.1.x) 7.0.11)" != "6.1.29" ]];  then
```
I'm not sure, what is this testing? Do we need to keep these version up-to-date when next 6.1/7.0 are released?

wadells replied:
Nothing in this file should need to be kept up to date for a couple reasons:

1) In order to avoid etcd mismatches the current algorithm will not recommend from tags/commits more current than the to build.  That means we should never recommend a CI upgrade from 6.1.40 -> 7.0.11.  It will be pinned at 6.1.29 forever.

2) These are unit tests for the code in utils.sh.  If the test cases are out of date, they will never affect what tests are actually run.  I'll add a comment to the file to help clarify what the intention is.

wadells replied:
Fixed in ec2a27a.

r0mant commented on build.assets/robotest/utils.sh:
```diff
@@ -0,0 +1,96 @@
+set -o pipefail
+
+function branch {
+  # GRAVITY_GIT_VERSION_BRANCH_PREFIX may defined to accommodate remote names other than"origin"
```
Nit: "may _be_ defined"? Also, space b/w "than" and "origin".

wadells replied:
"may" is intentional.  If `GRAVITY_GIT_VERSION_BRANCH_PREFIX` is undefined, `branch_prefix` defaults to `remotes/origin/version` which should be sane for CI and most developer setups.

This variable is to accommodate folks like me, who don't name their remotes 'origin':

```
walt@work:~/git/gravity$ git remote -v
grav    git@github.com:gravitational/gravity.git (fetch)
grav    git@github.com:gravitational/gravity.git (push)
walt    git@github.com:wadells/gravity.git (fetch)
walt    git@github.com:wadells/gravity.git (push)
```

wadells replied:
Whoops, this was a comment about the missing "be".  Will fix.

wadells replied:
Fixed in ec2a27a.

r0mant commented on build.assets/robotest/pr_config.sh:
```diff
@@ -0,0 +1,70 @@
+# UPGRADE_MAP[$(recommended_upgrade_tag $(branch 6.2.x))]="redhat:7" # compatible non-LTS version
+# UPGRADE_MAP[6.2.0]="ubuntu:16"
+
+UPGRADE_VERSIONS=${!UPGRADE_MAP[@]}
```
Can we dump the entire constructed upgrade matrix to CLI (unless we already do that)? Just as a log, to be able to see calculated versions in the job output, in case the algorithm malfunctions. Same for nightly config.

wadells replied:
We already do that:

https://github.com/gravitational/gravity/blob/d4b3c67b2c8f785883fb1581c36e5ab435a6f71f/build.assets/robotest_run_suite.sh#L111

The above was annoying to test though, as it required adding/remvoing an exit call to only view the config without executing it.

This changset makes the configs easier to check in an adhoc manner.  See following, from "Testing done":

<details><summary>./build.assets/robotest/pr_config.sh</summary>

```
walt@work:~/git/gravity$ ./build.assets/robotest/pr_config.sh     
install={"flavor":"three","nodes":3,"role":"node","os":"redhat:7","storage_driver":"overlay2"}
install={"installer_url":"/installer/opscenter.tar","nodes":1,"flavor":"standalone","role":"node","os":"ubuntu:18","ops_advertise_addr":"example.com:443"}
resize={"to":3,"flavor":"one","nodes":1,"role":"node","state_dir":"/var/lib/telekube","os":"ubuntu:18","storage_driver":"overlay2"}
shrink={"nodes":3,"flavor":"three","role":"node","os":"redhat:7"}
upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:16","from":"/telekube_7.0.0.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"flavor":"three","nodes":3,"role":"node","os":"centos:7","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"flavor":"three","nodes":3,"role":"node","os":"debian:9","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"flavor":"three","nodes":3,"role":"node","os":"ubuntu:18","from":"/telekube_7.0.11.tar","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
```
</details>

wadells commented on build.assets/robotest/test.sh:
```diff
@@ -0,0 +1,23 @@
+fi
+
+# ensure an upgrade from a release to itself is not recommended
+if [[ "$(recommended_upgrade_tag_between $(branch 7.0.x) 7.0.11)" != "7.0.10" ]];  then
```
I'm not sure what you mean by "Is 7.0.11 non-inclusive here?"

This test is to make the algorithm does not recommend a 7.0.11 -> 7.0.11 upgrade -- or worse a 7.0.12 -> 7.0.11 upgrade. 7.0.10 should be the recommended 7.0.x upgrade to 7.0.11.

Effectively, this parses as: "When upgrading to 7.0.11 from a recent 7.0.x build, the code should recommend an upgrade from 7.0.10."

The specific values 7.0.11 and 7.0.10 don't particularly matter.  The test "When upgrading to 6.1.5 from a recent 6.1.x, the code should recommend 6.1.4."

Let me know if I could clarify the code or comment somehow to get this idea across more cleanly.

wadells replied:
I added significant documentation to and about test.sh in 257f484.  I also moved this back to 7.0.10, just to further indicate that these aren't supposed to be updated/current.

Let me know if any further clarification would help.

wadells committed:
Add further comments to test cases.

These tests are confusing to reviewers:

  https://github.com/gravitational/gravity/pull/1837#discussion_r451781632
  https://github.com/gravitational/gravity/pull/1837#discussion_r452266434

Hopefully this documentation will help clarify why they exist.

wadells committed:
Remove spurrious OSes in PR upgrade matrix.

Per discussion here:

  https://github.com/gravitational/gravity/pull/1837#discussion_r451780198

These were accidentally added.

wadells committed:
Fix comment typos.

Per review at:

  https://github.com/gravitational/gravity/pull/1837#discussion_r451781830

a-palchikov commented on build.assets/robotest/test.sh:
```diff
@@ -0,0 +1,50 @@
+
+# Check recommended upgrade when patch-1  doesn't exist.
+#
+# 6.1.23 is the only recent case of this in the wild.
```
So, I'm still unsure how this works. If you wanted to skip `6.1.23` -> `6.1.24` upgrade path, why is it not excluding `6.1.23` explicitly and instead uses a condition to avoiding anything other than `6.1.22`?
I guess my question is - how is this going to be used and when would this error (eg `exit 1`) be triggered?

wadells replied:
> why is it not excluding 6.1.23 explicitly and instead uses a condition to avoiding anything other than 6.1.22?

Perhaps it would be easier to read if the condition wasn't negated?

```bash
if [[ "$(recommended_upgrade_tag_between $(branch 6.1.x) "6.1.24")" == "6.1.22" ]];  then
  echo "Test passed"
else
  exit 1
fi
```

Or are you suggesting:

```bash
if [[ "$(recommended_upgrade_tag_between $(branch 6.1.x) "6.1.24")" == "6.1.23" ]];  then
  exit 1
else
  echo "Test passed"
fi
```

The second test case allows for other silliness like recommending 6.1.24.  So while it is more pure, it is less useful as a test IMO.

> I guess my question is - how is this going to be used and when would this error (eg exit 1) be triggered?

These unit tests are used used by calling the file directly:

```
walt@work:~/git/gravity$ ./build.assets/robotest/test.sh 
++ dirname ./build.assets/robotest/test.sh    
+ source ./build.assets/robotest/utils.sh                                                      
++ set -o errexit                                                                                                                                                                             
++ set -o nounset                                                                              
++ set -o pipefail                 
+++ branch 6.1.x                                                                               
+++ local branch_prefix=remotes/grav/version
...
```

Exit 1 is triggered if any test cases fails.  I'd expect that following exit 1, the employee who triggered the test would look at why test is there and if recommended_upgrade_tag_between has changed.

This file is not presently integrated with `make test` or any make targets, so it won't be run by any CI.  I could probably clean it up (turn off `xtrace`, add some logging statements) and integrate it.  However my my goal I wasn't to invent or integrate a bash test harness.  I am trying to make sure useful builds were recommended in these situations, and writing the these tests seemed like the easiest way to do so.

As a second goal. these tests serve as concrete documenation for what the algorithm needs to handle.  If/when someone re-implements a better version of this (outside of bash? maybe ditching git tags as a strategy?), the scenarios these test cases check should still be relevant.

I'd considered moving to a better language myself, but go -> bash (robotest's `run_suite.sh`) -> go felt more byzantine than the current bash configuration, and I didn't have time/priority to translate all of robotest's bindings into native go.

wadells replied:
Because:

1) The only remaining discussion point is test code that does not affect execution
2) I've been iterating on this PR for 9 days

I'm going to go ahead and merge as is.

That being said, I do want to get to the bottom of this -- I'll just handle any additional changes in a follow up PR.

I see ~4 reasonable paths:

1) leave it as is (if I've convinced you in the discussion here)
2) refactor test cases to be more clear (see https://github.com/gravitational/gravity/pull/1837#discussion_r452948904)
3) option 2 + integrate tests with make/CI such that they're regularly run
4) remove the unit tests entirely.... They cause more confusion than they prevent.

I prefer one of 1-3, and I think 2 is closest to what you're looking for.
